File: packages/Api.NCore/Model/Authenticate.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Api.NCore.Utils;
using Api.NCore.Model.User;

namespace Api.NCore.Model
{
    public static class Authenticate
    {
        private static readonly HttpClient Client = new HttpClient();

        /**
         * Create state key, store in user database and return generated url for redirection
         */
        public static async Task<string> AuthenticateAccount(string id, string service, AuthOptions options)
        {
            var stateKey = Generate.GenerateRandomString(64);
            await UpdateTokens(id, service, new {StateKey = stateKey});
            var url = GenerateUrl(options, stateKey);
            return url;
        }

        /**
         * Method fired after redirect from service
         */
        public static async Task<bool> GetAccessToken(string id, string service, GetAccessTokenParams param)
        {
            var stateKey = await GetStateKey(id, service);
            if (param.State != stateKey || param.State == null) return false;

            // Update user tokens - Code and reset StateKey to ""
//            await UpdateTokens(id, service, new {param.Code, StateKey = ""});
            // Create form for post request
            var form = Generate.GenerateAuthForm(param);
            var req = await Client.SendAsync(form);

            Console.WriteLine("CONTENT = {0}", await req.Content.ReadAsStringAsync());
            return true;
        }

        public static string GenerateBasicAuthorization(string clientId, string clientSecret)
        {
            var encode =
                Convert.ToBase64String(Encoding.ASCII.GetBytes(string.Format("{0}:{1}", clientId, clientSecret)));
            return $"Basic {encode}";
        }

        /**
         * Update tokens
         */
        private static async Task UpdateTokens(string id, string serviceName, object tokens)
        {
            var service = new Dictionary<string, object> {[serviceName] = tokens};
            await UserManager.UpdateCredentials(id, new
            {
                AuthTokens = service
            });
        }

        /**
         * Return user state key
         */
        private static async Task<string> GetStateKey(string id, string serviceName)
        {
            var key = await UserManager.GetUser(id);
            return key.AuthTokens[serviceName].StateKey;
        }

        /**
         * Generate url for redirect
         */
        private static string GenerateUrl(AuthOptions options, string stateKey)
        {
            return
                $"{options.Url}?client_id={options.ClientId}&response_type=code&scope={Uri.EscapeUriString(options.Scopes)}&redirect_uri={Uri.EscapeUriString(options.Redirect)}{(options.State ? "&state=" + stateKey : "")}{(options.Nonce ? "&nonce=" + stateKey : "")}";
        }
    }
}

File: packages/Api.Node/app/controller/Authenticate.ts
import request from "request";
import signale from "signale";
import { GetUser, UpdateTokens } from "#/components/user/Manager";
import { IAccessTokenParams, IAuthenticationParams, IRefreshToken } from "#SH/Interfaces";
import { IAuthTokens } from "#SH/Interfaces";
import { generateRandomString } from "#SH/Utils";

/**
 * Generate basic authorization
 */
const GenerateBasicAuthorization = (clientID: string, clientSecret: string): string => {
  return `Basic ${Buffer.from(clientID + ":" + clientSecret).toString("base64")}`;
};

export const TokensNotExists = (service: string, tokens: IAuthTokens): boolean => {
  return tokens.AuthTokens[service] === undefined || tokens.AuthTokens[service].AccessToken === undefined;
};

/**
 * Receive new refresh token from service
 */
export const RefreshTokens = async ({ id, service, url, auth }: IRefreshToken): Promise<boolean> => {
  try {
    const basicAuth = GenerateBasicAuthorization(auth.clientID, auth.clientSecret);
    const user: IAuthTokens = await GetUser(id);

    // If service doesn't exists stop requesting refresh token
    if (user.AuthTokens[service] === undefined) {
      return false;
    }

    const refreshToken = user.AuthTokens[service].RefreshToken;
    const authOptions = {
      url,
      form: {
        grant_type: "refresh_token",
        refresh_token: refreshToken
      },
      headers: {
        Authorization: basicAuth
      },
      json: true
    };

    request.post(authOptions, async (error: any, response: request.Response, body: any) => {
      if (!error && response.statusCode === 200) {
        const AccessToken = body.access_token;

        await UpdateTokens({
          id,
          service,
          tokens: { AccessToken }
        });
      }
    });
    return true;
  } catch (e) {
    signale.error("Authenticate.RefreshToken ------", e);
    throw Error(e);
  }
};

export default class Authenticate {
  protected id: string = "";
  protected service: string = "";

  /**
   * Generate authentication url with needed state key
   */
  public async AuthenticateAccount(id: string, service: string, options: IAuthenticationParams): Promise<string> {
    try {
      this.id = id;
      this.service = service;
      // Generate state key for service authentication to ensure that result code is from service
      const StateKey = generateRandomString(64);
      // Add state key to user account
      await UpdateTokens({
        id: this.id,
        service: this.service,
        tokens: { StateKey }
      });

      return (
        `${options.url}?client_id=${options.clientID}` +
        `&response_type=code` +
        `&scope=${encodeURI(options.scopes)}` +
        `&redirect_uri=${encodeURI(options.redirect_uri)}` +
        `&state=${options.state || StateKey}` +
        `${options.nonce ? "&nonce" + options.nonce : ""}`
      );
    } catch (e) {
      signale.error("Authenticate.AuthenticateAccount ------", e);
      throw Error(e);
    }
  }

  /**
   * Send request to service in order to receive access and refresh tokens
   */
  public async GetAccessToken({ code, state, Authorization, url, redirect_uri }: IAccessTokenParams): Promise<boolean> {
    try {
      const stateKey = await this.GetStateKey();
      // If state key from request is different from one stored in db
      // Return false and end authentication
      if (state !== stateKey || state === null) {
        return false;
      } else {
        // Reset state key and update code
        await UpdateTokens({
          id: this.id,
          service: this.service,
          tokens: {
            Code: code,
            StateKey: ""
          }
        });

        const basicAuth = GenerateBasicAuthorization(Authorization.clientID, Authorization.clientSecret);

        const form = {
          url,
          form: {
            code,
            grant_type: "authorization_code",
            ...(redirect_uri !== undefined ? { redirect_uri } : "")
          },
          headers: {
            Authorization: basicAuth
          },
          json: true
        };

        // Send request to service to receive access token and refresh token
        request.post(form, async (error: any, response: request.Response, body: any) => {
          if (!error && response.statusCode === 200) {
            const AccessToken = body.access_token;
            const RefreshToken = body.refresh_token;
            const ExpiresIn = body.expires_in;
            const Scope = body.scope;
            // Store tokens in database
            await UpdateTokens({
              id: this.id,
              service: this.service,
              tokens: {
                ...(AccessToken !== undefined ? { AccessToken } : ""),
                ...(RefreshToken !== undefined ? { RefreshToken } : ""),
                ...(ExpiresIn !== undefined ? { ExpiresIn } : ""),
                ...(Scope !== undefined ? { Scope } : "")
              }
            });
          }
        });
        return true;
      }
    } catch (e) {
      signale.error("Authenticate.GetAccessToken ------", e);
      throw Error(e);
    }
  }

  /**
   * Get state key from database
   */
  protected async GetStateKey(): Promise<string> {
    try {
      const user: IAuthTokens = await GetUser(this.id);
      return user.AuthTokens[this.service].StateKey;
    } catch (e) {
      signale.error("Authenticate.GetStateKey ------", e);
      throw Error(e);
    }
  }
}
